### DISCUSSION TITLE
Optimize rendering of common objects

### DISCUSSION THREAD

**Discussion by zaalbarxx** [fabricjs/fabric.js#8587] [Q&A] [answered]
Hi!
I am working on some kind of seating plan application which looks like this
![image](https://user-images.githubusercontent.com/3301242/211572761-22531c57-f68c-475a-a793-3168b728a50c.png)
![image](https://user-images.githubusercontent.com/3301242/211572870-c2fd402e-5656-45ce-bc21-ff3e8c78f7c7.png)

Generally speaking everything is working, but there are some performance issues with such big plans (which can include up to 3000 seats which can translate to around 3000 `Group` objects and 12-15k objects (4-5 per group - seat).

 I was wondering if there are any tricks that could be used here to improve the performance since a lot of these "seats" look very similar (like 1000 can be of the same color). So I was thinking if some optimization could be done here to for example have some shared cache for these - so all 1000 objects would share same cached object for rendering. Or if there are maybe other things that could be done here. 

The biggest problem here seem to be the zooming (very clunky) and moving/panning. Interesingly with bigger zoom the app works better which for me is weird since I would guess that the total canvas dimensions in that scenario will be bigger.
The other problem is definitely the memory consumption which as we found out is so big on Safari that it crashes the browser making the app totally unusable.

Do you guys have any advices what could I look at to maybe make it more responsive and less memory hungry ? I was playing around with pixi.js where there is a concept of textures and with approach of making 1 texture per seat color and re-using these textures the app worked in almost 60fps, the con was unfortunately that it looked blurry and texts were looking badly, but I guess that would be something that could be worked out. What do you think ? Or maybe I should take a look at some other library for such case (recently I was thinking about Leaflet JS) ?

**Comment by ShaMan123** [comment, marked as answer]
[brief](https://github.com/fabricjs/fabric.js/issues/8581#issuecomment-1376601438)
[immediate workaround](https://github.com/fabricjs/fabric.js/discussions/7874#discussioncomment-2541974) and [source code](https://github.com/fabricjs/fabricjs.com/blob/b73f0811cc09a331acf2326e6a852a5dead25b13/posts/benchmarks/_posts/2011-9-1-raphael-vs-fabric-simple.html#L96-L126)

**Reply by ShaMan123** [reply]
> The biggest problem here seem to be the zooming (very clunky) and moving/panning. Interesingly with bigger zoom the app works better which for me is weird since I would guess that the total canvas dimensions in that scenario will be bigger.

This is due to an optimization that checks if objects are on screen before rendering them

**Reply by ShaMan123** [reply]
try to play with the group caching as well, not sure what would be better, on or off

**Reply by ShaMan123** [reply]
There is another complementary approach, sort of sharding
#7871

**Reply by ShaMan123** [reply]
I think I would render the common cache black and then apply a mask over it with the color of the seat with a globalCompositeOperation of the seat's color

**Reply by ShaMan123** [reply]
Or alternatively have a cache for each color, which might be more performant even

**Reply by zaalbarxx** [reply]
Yeah, that's what I was also thinking about (caching per seat color), but I could not find any informations about how to share cache between objects or create some "shared" object/texture. Basically 80% of this app is kind of "same" so I can only imagine that there could be some way to create like 30-40 shared objects which could then be quickly re-drawn in multiple places with GPU.

**Reply by zaalbarxx** [reply]
Besides of that there is still a very big issue with Safari running out of memory with big number of objects, is there any way to kind of "defer" rendering to batches so maybe that would cause not all objects to immediately fill the memory (I've read that it has some limit on GPU allocated memory) and thus maybe not crash ? This is not observed in Chrome and I suppose it is due to the fact that Chrome is far more liberal when it comes to memory usage :)

**Reply by ShaMan123** [reply]
All the above will save you from safari OOM
objectCaching defaulting to true is a really bad thing for big apps.
These apps must manage it more closely

**Reply by zaalbarxx** [reply]
So what you suggest is to set `objectCaching: false` for all objects or only for groups (each group contain seat shape and text/flag/other stuff) ? And I assume that there is not much I can do here with sharing the cache before the #8581 will make it?

**Reply by ShaMan123** [reply]
> These apps must manage it more closely

This is my suggestion, the entire thread.
Go for the immediate workaround, definitely.

**Reply by ShaMan123** [reply]
Or more straight forward, feel the difference:
[with workaround](http://fabricjs.com/raphael-vs-fabric-simple?n=5000&optimize_caching=1)
[**without** workaround](http://fabricjs.com/raphael-vs-fabric-simple?n=5000&optimize_caching=0)

select and drag elements

**Comment by ShaMan123** [comment]
Ohh part of what I referenced points to the layering approach.
Basically you put all non active objects into a group with caching so when the canvas renders the group is rendered once and as long as it doesn't change each rendering it has zero effect on perf since it renders the cache.
Up to you to decide how many groups you want to create and logic behind that.
#7871 is a POC

**Reply by ShaMan123** [reply]
Ahh I see I wrote that already
